sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEPropagateEmptyRelation.scala

```diff
@@ -53,8 +54,8 @@ object AQEPropagateEmptyRelation extends PropagateEmptyRelationBase {
       empty(j)
   }
 
-  // TODO we need use transformUpWithPruning instead of transformUp
-  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUp {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
+    _.containsAnyPattern(LOGICAL_QUERY_STAGE, LOCAL_RELATION, TRUE_OR_FALSE_LITERAL), ruleId) {
```

Comment:
Can AQEPropagateEmptyRelation be invoked multiple times on the same tree or subtree?  
If not, then we don't need "ruleId".

Reply:
This rule is in a "fixedPoint" batch, and can be run multiple rounds.

Reply:
Ah, make sense. I somehow read it as in a run-once batch.

Reply:
yes, in the latest version the rule used `fixedPoint` batch.